File: src/realm.js
/**
 * Physical vs digital realm classification.
 * Physical = human body / offline world must act.
 * Digital = agents/swarm can advance (subject to privacy/HITL).
 */

/** @typedef {'physical' | 'digital'} Realm */

const PHYSICAL_KINDS = new Set([
  'physical_errand',
  'physical_presence',
  'outdoor',
  'caregiving_in_person',
  'health_body',
  'household',
]);

const PHYSICAL_KEYWORDS = [
  /\boutdoor\b/i,
  /\bwalk\b/i,
  /\bgrocer/i,
  /\berrand\b/i,
  /\bpick\s*up\b/i,
  /\bfamily\s+support\b/i,
  /\bplay\s+with\b/i,
  /\bnaps?\b/i,
  /\blight\s+(weight\s+)?lift/i,
  /\bgarden/i,
  /\bin[- ]person\b/i,
  /\bcommute\b/i,
  /\bcook\b|\bmeal\s+prep\b/i,
  /\bput\s+(the\s+)?(baby|son|child)\s+to\s+sleep\b/i,
  /\bpresence\s+block\b/i,
  /\bwind-?down\b/i,
  /\bsleep\b/i,
  /\bbreakfast\b|\blunch\b|\bdinner\b/i,
];

const DIGITAL_KEYWORDS = [
  /\bgithub\b|\bpr\b|\bcommit\b|\boss\b/i,
  /\bemail\b|\bapply\b|\bapplication\b/i,
  /\bcv\b|\bportfolio\b|\bdevprofile\b/i,
  /\bcode\b|\bship\b|\bdeploy\b|\brefactor\b/i,
  /\bswarm\b|\bagent\b|\bcli\b/i,
];

/**
 * @param {{ id?: string, title?: string, area?: string, kind?: string, tags?: string[], realm?: string, body?: string, non_negotiable?: boolean, capacity?: string }} item
 * @returns {{ realm: Realm, reason: string }}
 */
export function classifyRealm(item = {}) {
  if (item.realm === 'physical' || item.realm === 'digital') {
    return { realm: item.realm, reason: `explicit realm:${item.realm}` };
  }

  const tags = (item.tags || []).map((t) => String(t).toLowerCase());
  if (tags.includes('physical') || tags.includes('offline') || tags.includes('errand')) {
    return { realm: 'physical', reason: 'tag physical/offline/errand' };
  }
  if (tags.includes('digital') || tags.includes('online')) {
    return { realm: 'digital', reason: 'tag digital/online' };
  }

  const kind = item.kind || '';
  if (PHYSICAL_KINDS.has(kind)) {
    return { realm: 'physical', reason: `kind ${kind}` };
  }

  const text = `${item.title || ''} ${item.body || ''} ${item.id || ''}`;
  for (const re of PHYSICAL_KEYWORDS) {
    if (re.test(text)) {
      return { realm: 'physical', reason: `matched ${re}` };
    }
  }

  // Relationship non-negotiable blocks and pure Health body blocks default physical
  if (item.non_negotiable && (item.area === 'Relationships' || item.area === 'Health')) {
    return { realm: 'physical', reason: 'non-negotiable presence/health block' };
  }

  for (const re of DIGITAL_KEYWORDS) {
    if (re.test(text)) {
      return { realm: 'digital', reason: `matched digital ${re}` };
    }
  }

  // Career/Systems/Learning default digital; Relationships/Health default physical when ambiguous
  if (item.area === 'Relationships' || item.area === 'Health') {
    return { realm: 'physical', reason: `area ${item.area} defaults physical when ambiguous` };
  }

  return { realm: 'digital', reason: 'default digital (agent-automatable)' };
}

/**
 * Attach realm classification to items.
 * @param {Array<object>} items
 */
export function enrichWithRealm(items = []) {
  return items.map((item) => {
    const realmInfo = classifyRealm(item);
    return { ...item, realm: realmInfo.realm, realmInfo };
  });
}

/**
 * Physical pickups the human must do (pending / not done).
 * @param {Array<object>} actions
 * @param {{ includeDone?: boolean }} opts
 */
export function physicalPickups(actions = [], opts = {}) {
  const enriched = enrichWithRealm(actions);
  return enriched.filter((a) => {
    if (a.realm !== 'physical') return false;
    if (!opts.includeDone && (a.status === 'done' || a.status === 'completed')) return false;
    return true;
  });
}

/**
 * Digital actions agents can progress (still may need HITL).
 */
export function digitalActions(actions = []) {
  return enrichWithRealm(actions).filter((a) => a.realm === 'digital');
}

export { PHYSICAL_KINDS };

File: crates/ensembly-kernel/src/realm.rs
//! Physical vs digital realm classification.
//! Port of ensembly `src/realm.js` core semantics.

use regex::Regex;
use serde::{Deserialize, Serialize};
use std::sync::OnceLock;

use crate::privacy::Classifiable;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Realm {
    Physical,
    Digital,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RealmInfo {
    pub realm: Realm,
    pub reason: String,
}

fn physical_kinds() -> &'static [&'static str] {
    &[
        "physical_errand",
        "physical_presence",
        "outdoor",
        "caregiving_in_person",
        "health_body",
        "household",
    ]
}

fn physical_res() -> &'static [Regex] {
    static RE: OnceLock<Vec<Regex>> = OnceLock::new();
    RE.get_or_init(|| {
        [
            r"(?i)\boutdoor\b",
            r"(?i)\bwalk\b",
            r"(?i)\bgrocer",
            r"(?i)\berrand\b",
            r"(?i)\bpick\s*up\b",
            r"(?i)\bin[- ]person\b",
            r"(?i)\bcook\b|\bmeal\s+prep\b",
        ]
        .into_iter()
        .map(|p| Regex::new(p).expect("physical regex"))
        .collect()
    })
}

/// Classify realm for an item.
pub fn classify_realm(item: &Classifiable) -> RealmInfo {
    // explicit realm via kind tags is preferred when encoded in tags
    if item.tags.iter().any(|t| t.eq_ignore_ascii_case("physical")) {
        return RealmInfo {
            realm: Realm::Physical,
            reason: "tag physical".into(),
        };
    }
    if item.tags.iter().any(|t| t.eq_ignore_ascii_case("digital")) {
        return RealmInfo {
            realm: Realm::Digital,
            reason: "tag digital".into(),
        };
    }

    let kind = item.kind.as_deref().unwrap_or("");
    if physical_kinds().contains(&kind) {
        return RealmInfo {
            realm: Realm::Physical,
            reason: format!("kind {kind}"),
        };
    }

    let text = format!(
        "{} {} {}",
        item.title.as_deref().unwrap_or(""),
        item.body.as_deref().unwrap_or(""),
        item.id.as_deref().unwrap_or("")
    );
    for re in physical_res() {
        if re.is_match(&text) {
            return RealmInfo {
                realm: Realm::Physical,
                reason: format!("matched {}", re.as_str()),
            };
        }
    }

    let area = item.area.as_deref().unwrap_or("");
    if area == "Relationships" || area == "Health" {
        return RealmInfo {
            realm: Realm::Physical,
            reason: format!("area {area} defaults physical when ambiguous"),
        };
    }

    RealmInfo {
        realm: Realm::Digital,
        reason: "default digital (agent-automatable)".into(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn errand_is_physical() {
        let r = classify_realm(&Classifiable {
            id: Some("grocery-errand".into()),
            title: Some("Grocery errand".into()),
            kind: Some("physical_errand".into()),
            ..Default::default()
        });
        assert_eq!(r.realm, Realm::Physical);
    }

    #[test]
    fn career_default_digital() {
        let r = classify_realm(&Classifiable {
            title: Some("Prepare application".into()),
            area: Some("Career".into()),
            kind: Some("job_application_submit".into()),
            ..Default::default()
        });
        assert_eq!(r.realm, Realm::Digital);
    }
}
